File: exts/jphp-gui-game-ext/src/main/java/org/develnext/jphp/ext/game/classes/UXSprite.java
package org.develnext.jphp.ext.game.classes;

import javafx.scene.canvas.Canvas;
import javafx.scene.image.Image;
import org.develnext.jphp.ext.game.GameExtension;
import org.develnext.jphp.ext.game.support.Sprite;
import php.runtime.annotation.Reflection.*;
import php.runtime.env.Environment;
import php.runtime.lang.BaseWrapper;
import php.runtime.reflection.ClassEntity;

@Namespace(GameExtension.NS)
public class UXSprite extends BaseWrapper<Sprite> {
    interface WrappedInterface {
        @Property double[] frameSize();
        @Property double frameWidth();
        @Property double frameHeight();
        @Property int speed();
        @Property Image image();
        @Property boolean cycledAnimation();
        @Property @Nullable String currentAnimation();

        void setAnimation(String name, int... indexes);
        void setAnimationSpeed(String name, int speed);

        void draw(Canvas node, int index);
        void drawByTime(Canvas canvas, long now);
        void drawNext(Canvas canvas);
        Image getFrameImage(int index);

        void freeze();
        void unfreeze();
        boolean isFreeze();

        void play(String animation);
        void play(String animation, int speed);
    }

    public UXSprite(Environment env, Sprite wrappedObject) {
        super(env, wrappedObject);
    }

    public UXSprite(Environment env, ClassEntity clazz) {
        super(env, clazz);
    }

    @Signature
    public void __construct() {
        __construct(null);
    }

    @Signature
    public void __construct(@Nullable Sprite sprite) {
        __wrappedObject = new Sprite(sprite);
    }

    @Getter
    public int getFrameCount() {
        return getWrappedObject().getMaxIndex() + 1;
    }
}

File: exts/jphp-gui-game-ext/src/main/resources/JPHP-INF/sdk/php/game/UXSprite.php
<?php
namespace php\game;

use php\gui\UXCanvas;
use php\gui\UXImage;
use php\lang\IllegalArgumentException;

/**
 * Class UXSprite
 * @package php\game
 * @packages game, javafx
 */
class UXSprite
{
    /**
     * @var UXImage
     */
    public $image;

    /**
     * @readonly
     * @var double[] width + height
     */
    public $frameSize = [0, 0];

    /**
     * @var double
     */
    public $frameWidth = 0;

    /**
     * @var double
     */
    public $frameHeight = 0;

    /**
     * @readonly
     * @var int
     */
    public $frameCount = 0;

    /**
     * @var int
     */
    public $speed = 12;

    /**
     * @var bool
     */
    public $cycledAnimation = true;

    /**
     * @var null|string
     */
    public $currentAnimation = null;

    /**
     * @param UXSprite $origin
     */
    public function __construct(UXSprite $origin = null)
    {
    }

    /**
     * @param string $name
     * @param int[] $indexes
     */
    public function setAnimation($name, array $indexes)
    {
    }

    /**
     * @param string $animation
     * @param int $speed (optional)
     */
    public function play($animation, $speed = -1)
    {
    }

    /**
     * @param UXCanvas $canvas
     * @param $index
     */
    public function draw(UXCanvas $canvas, $index)
    {
    }

    /**
     * @param UXCanvas $canvas
     */
    public function drawNext(UXCanvas $canvas)
    {
    }

    /**
     * @param UXCanvas $canvas
     * @param int $now
     */
    public function drawByTime(UXCanvas $canvas, $now)
    {
    }

    /**
     * @param int $index
     * @return UXImage
     * @throws IllegalArgumentException
     */
    public function getFrameImage($index)
    {
    }
}